Add command import node store to shards

miguelportilla opened the pull request.

ripplelabs-jenkins commented:
## Jenkins Build Summary

Built from [this commit](https://github.com/ripple/rippled/commit/cd08ceba034ff6fac1c5db7cda36c6a435a63890)

Built at __20180319 - 22:49:33__

### Test Results

Build Type | Log | Result | Status
---------- | --- | ------ | ------
docs | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2441/1//artifact/docs.txt) | 1 cases, 0 failed, t: 2s | PASS :white_check_mark: 
clang.debug.unity | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2441/1//artifact/clang.debug.unity.txt) | 1003 cases, 0 failed, t: 245s | PASS :white_check_mark: 
msvc.debug | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2441/1//artifact/msvc.debug.txt) | 999 cases, 0 failed, t: 554s | PASS :white_check_mark: 
coverage | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2441/1//artifact/coverage.txt) | 1003 cases, 0 failed, t: 599s | PASS :white_check_mark: 
clang.debug.unity, <br/>PARALLEL_TESTS=false | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2441/1//artifact/clang.debug.unity_PARALLEL_TESTS_false.txt) | 1003 cases, 0 failed, t: 467s | PASS :white_check_mark: 
msvc.debug, <br/>PROJECT_NAME=rippled_classic | [console](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2441/1//console) | no test results, t: n/a | FAIL :red_circle: 
clang.debug.nounity | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2441/1//artifact/clang.debug.nounity.txt) | no test results, t: 297s | FAIL :red_circle: 
gcc.debug.unity | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2441/1//artifact/gcc.debug.unity.txt) | 1003 cases, 0 failed, t: 235s | PASS :white_check_mark: 
gcc.debug.nounity | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2441/1//artifact/gcc.debug.nounity.txt) | no test results, t: 302s | FAIL :red_circle: 
msvc.release | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2441/1//artifact/msvc.release.txt) | 998 cases, 0 failed, t: 496s | PASS :white_check_mark: 
clang.release.unity | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2441/1//artifact/clang.release.unity.txt) | 1002 cases, 0 failed, t: 325s | PASS :white_check_mark: 
gcc.release.unity | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2441/1//artifact/gcc.release.unity.txt) | 1002 cases, 0 failed, t: 343s | PASS :white_check_mark: 
gcc.debug.unity <br/>-Dstatic=true | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2441/1//artifact/gcc.debug.unity_Dstatic_true.txt) | 1003 cases, 0 failed, t: 235s | PASS :white_check_mark:

bachase commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -748,6 +921,94 @@ DatabaseShardImp::selectCache(std::uint32_t seq)
 }
 
+bool
+DatabaseShardImp::importCopy(std::unique_lock<std::mutex>&,
```
This looks a lot like `DatabaseShardImp::copyLedger`. Are they too different to unify?

miguelportilla replied:
This one lacks some checks that are unnecessary here. It also doesn't lock and make changes to the `complete_` and `incomplete_` members. Its possible to unify but not sure if it warrants the effort.

miguelportilla replied:
I forgot to mention this version does not pollute the shard's caches, a performance improvement.

bachase commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -333,6 +345,167 @@ DatabaseShardImp::validate()
+            return;
+        }
+        earliestIndex = seqToShardIndex(seq);
+        if (seq != firstLedgerSeq(earliestIndex))
```
While diving into this code, I found an oddity in the assert and subsequent line at https://github.com/ripple/rippled/blob/cd08ceba034ff6fac1c5db7cda36c6a435a63890/src/ripple/nodestore/impl/DatabaseShardImp.h#L88-L89

I'm not clear on how what if clause is checking.

miguelportilla replied:
The first ledger of a truncated shard. eg. XRP Genesis or alt using some other earliest truncated shard.

mellery451 commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -748,6 +921,94 @@ DatabaseShardImp::selectCache(std::uint32_t seq)
+    if (!incomplete_->setStored(ledger))
+        return false;
+    auto const after {incomplete_->fileSize()};
+     if(after > before)
```
nit extra space indent

seelabs commented:
@mellery451 This PR in 1.0.0-b2 should fix that warning: https://github.com/ripple/rippled/pull/2422

bachase commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -333,6 +345,167 @@ DatabaseShardImp::validate()
+            auto const numLedgers {shardIndex == earliestShardIndex()
+                ? lastSeq - firstSeq + 1 : ledgersPerShard_};
+            auto ledgerHashes{getHashesByIndex(firstSeq, lastSeq, app_)};
+            if (ledgerHashes.size() != numLedgers)
```
I'd consider adding some logging. I just ran the command on a node store that has fairly fragmented history. It initially reports `Importing 236 shards from node store` but there aren't actually any shards in that range that are complete.  

Alternatively, you could defer the `Importing ...` message until after you do a pass to check whether the hashes are available for all ledgers in each shard.

miguelportilla replied:
This is not an error and should not log. There simply isn't enough ledgers in the node store to complete this shard so it moves onto the next. However the importing message does need to be moved.

HowardHinnant commented on src/ripple/nodestore/impl/Shard.h:
```diff
@@ -37,7 +37,7 @@ namespace NodeStore {
 using PCache = TaggedCache<uint256, NodeObject>;
```
Missing `#include <ripple/nodestore/DatabaseShard.h>`.  Nonunity won't build without it.

miguelportilla commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -333,6 +345,167 @@ DatabaseShardImp::validate()
+        }
+    }
+
+    std::cout << "Importing " << (latestIndex - earliestIndex + 1) <<
```
I can change this to a log, but if the user's severity is lower, they won't receive feedback. Assuming severity, it takes a while before log messages appear as this is a lengthy and time consuming operation. I think some feedback is helpful. Open to suggestions.

miguelportilla replied:
Moved to after a successful copy and changed to log at debug severity.

miguelportilla commented on src/ripple/app/main/Application.cpp:
```diff
@@ -2064,12 +2068,40 @@ bool ApplicationImp::updateTables ()
+        Throw<std::runtime_error>("no overlay");
+    if(config_->standalone())
+    {
+        // node import requires the sqlite db
```
Not stale

miguelportilla commented:
@mellery451 @bachase thanks for the feedback and thorough review. I believe I addressed all of the comments, please let me know if I missed something.

bachase commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -415,15 +421,13 @@ DatabaseShardImp::importNodeStore()
-
-        // Skip if directory exists
+        if (complete_.find(shardIndex) != complete_.end() ||
+            (!incomplete_ && incomplete_->index() == shardIndex))
```
Should `!incomplete` be `incomplete`?

miguelportilla replied:
Yes! thanks.

bachase commented on src/ripple/nodestore/impl/DatabaseShardImp.cpp:
```diff
@@ -323,6 +345,176 @@ DatabaseShardImp::validate()
+        {
+            std::shared_ptr<Ledger> ledger;
+            std::uint32_t seq;
+            std::tie(ledger, seq, std::ignore) = loadLedgerHelper(
```
I'm getting a crash where this call to `loadLedgerHelper` ends up trying to relock the mutex `m_`. The call stack is 

```
        rippled.exe!ripple::NodeStore::DatabaseShardImp::contains(unsigned int seq) Line 295	C++
 	rippled.exe!ripple::NodeStore::DatabaseShardImp::fetchLedger(const ripple::base_uint<256,void> & hash, unsigned int seq) Line 209	C++
 	rippled.exe!ripple::InboundLedger::init(std::unique_lock<std::recursive_mutex> & collectionLock) Line 116	C++
 	rippled.exe!ripple::InboundLedgersImp::acquire(const ripple::base_uint<256,void> & hash, unsigned int seq, ripple::InboundLedger::Reason reason) Line 94	C++
 	rippled.exe!ripple::detail::AppFamily::acquire(const ripple::base_uint<256,void> & hash, unsigned int seq) Line 100	C++
 	rippled.exe!ripple::Ledger::Ledger(const ripple::LedgerInfo & info, bool & loaded, const ripple::Config & config, ripple::Family & family, beast::Journal j) Line 265	C++
 	[External Code]	
 	rippled.exe!ripple::loadLedgerHelper(const std::basic_string<char,std::char_traits<char>,std::allocator<char> > & sqlSuffix, ripple::Application & app) Line 1176	C++
 	rippled.exe!ripple::NodeStore::DatabaseShardImp::importNodeStore::__l3::<lambda>(bool ascendSort) Line 361	C++
```

miguelportilla replied:
I see the problem, thanks.

miguelportilla commented:
@mellery451 @bachase The last commit unifies the `copyledger` function. I moved it to the base class which reduces four copies in the derived classes. It also uses batch writes for increased performance. Please test and let me know if you encounter any issues, thanks!
